---
title: title
author: author
institute: place
format: html
execute:
  cache: true
---

# Lib
```{r}
library(data.table)
#library(rmarkdown)
library(SeqArray)
library(SNPRelate)
```

# Data
Ancestry-level (multi-GWAS) summary stats
```{r}
#| cache: true
eas <- fread("../data/DIAMANTE2022/sumstat/DIAMANTE-EAS.sumstat.txt", header=T)
eur <- fread("../data/DIAMANTE2022/sumstat/DIAMANTE-EUR.sumstat.txt", header=T)
sas <- fread("../data/DIAMANTE2022/sumstat/DIAMANTE-SAS.sumstat.txt", header=T)
```

Start with GCKR locus, index SNV rs1260326 (from supplementary table 7)\
```{r}
index_snv <- "rs1260326"
index_snv_pos <- eas[rsID == index_snv, `position(b37)`]
# TODO: In this case, this SNV's position is the same in all ancestries
# But write out generalizable form

eas_gckr_snvs <- eas[`position(b37)` < index_snv_pos+100000 &
                     `position(b37)` > index_snv_pos-100000  ]
eur_gckr_snvs <- eur[`position(b37)` < index_snv_pos+100000 &
                     `position(b37)` > index_snv_pos-100000  ]
sas_gckr_snvs <- sas[`position(b37)` < index_snv_pos+100000 &
                     `position(b37)` > index_snv_pos-100000  ]
# TODO: replace this w/ GDS seqSet 
```

Subset to SNPs shared by all ancestries, and sort
```{r}
# Very fast but could replace this with GDS file stuff too
# TODO: fast yes, but loading in eas/eur/sas ~beforehand~ is not.
common_snvs <- Reduce(intersect, list(eas_gckr_snvs$rsID,eur_gckr_snvs$rsID,sas_gckr_snvs$rsID))
eas_gckr_snvs <- eas_gckr_snvs[rsID %in% common_snvs]
eur_gckr_snvs <- eur_gckr_snvs[rsID %in% common_snvs]
sas_gckr_snvs <- sas_gckr_snvs[rsID %in% common_snvs]

eas_gckr_snvs <- eas_gckr_snvs[order(rsID)]
eur_gckr_snvs <- eur_gckr_snvs[order(rsID)]
sas_gckr_snvs <- sas_gckr_snvs[order(rsID)]
```

# ieugwasr (WIP)
nvm, uses VCF, old
```{r}
#library(ieugwasr)
#ld_reflookup("rs1260326", pop="EAS")
```

# SNPRelate (WIP)
```{r}
#gds_eas <- snpgdsOpen("../data/ref/1kg/eas/g1000_eas.gds")
#gds_eur <- snpgdsOpen("../data/ref/1kg/eur/g1000_eur.gds")
#gds_sas <- snpgdsOpen("../data/ref/1kg/sas/g1000_sas.gds")

# TODO: which correlation method is preferred?
# TODO: ERROR: some snp.id do not exist!
#snpgdsLDMat(gds_eas, snp.id=common_snvs, num.thread=NA)
#snpgdsLDMat(gds_eur, snp.id=common_snvs, num.thread=NA)
#snpgdsLDMat(gds_sas, snp.id=common_snvs, num.thread=NA)
```

```{r}
gds_1kg <- seqOpen("../data/ref/1kg/gds_format/1KG_ALL.autosomes.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.gds")
sample_info <- fread("../data/ref/1kg/gds_format/integrated_call_samples_v3.20130502.ALL.panel", header=F)

samples_eas <- sample_info[V3 == "EAS", V1]
samples_eur <- sample_info[V3 == "EUR", V1]
samples_sas <- sample_info[V3 == "SAS", V1]

snv_inds <- seqSetFilterAnnotID(gds_1kg, common_snvs, ret.idx=T)
# seqFilterPush(gds_1kg) # No need to push, variant filter not overwritten by the following filters on samples

seqSetFilter(gds_1kg, sample.id=samples_eas)
eas_inds <- seqGetFilter(gds_1kg)
seqSetFilter(gds_1kg, sample.id=samples_eur)
eur_inds <- seqGetFilter(gds_1kg)
seqSetFilter(gds_1kg, sample.id=samples_sas)
sas_inds <- seqGetFilter(gds_1kg)

identical(eas_inds$variant.sel, eur_inds$variant.sel, sas_inds$variant.sel)

ld_eas <- snpgdsLDMat(gds_1kg, snp.id=eas_inds$

#seqSummary(gds_1kg)
#seqSetFilter(gds_1kg, variant.id=common_snvs) # filters out e/t
#seqResetFilter(gds_1kg)
#head(seqGetData(gds_1kg, "annotation/id"))
ld_1kg <- snpgdsLDMat(gds_1kg, snp.id=snv_inds, num.thread=3) # note: sliding window defaults to 250
fwrite(ld_1kg$LD, file="1kg_ld_DIAMANTE_common_snvs-ALL.ld", sep=" ")
```





